File: src/conductances/Nap_Et2.py
"""
Nap_Et2: Persistent sodium conductance

Extracted from BBP NEURON mechanism: Nap_Et2.mod
Neuron type: pyramidal neuron

State variables:
    m: activation gate
    h: inactivation gate
"""

# temperature correction
QT = 2.3**((34 - 21) / 10)

# euler's constant to 30 digits of precision
e = 2.718281828459045235360287471352

# maximal peak conductance density of channel per unit membrane area (S/cm^2)
gNap_Et2bar = 0.00001

def derivatives(state, V):
    """
    Compute steady-state values and time constants,
    then compute time derivatives of the gating variables.

    Inputs:
        - state (tuple of floats): (m, h) gating variables
        - V (float): membrane potential (mV)

    Returns: 
        - dm_dt (float): time derivative of m
        - dh_dt (float): float time derivative of h
    """

    # steady state activation
    mInf = 1.0 / (1.0 + e**((V + 52.6) / -4.6))

    # voltage shift
    if V == -38:
        V += 0.0001

    # activation time constant
    mAlpha = (0.182 * (V + 38)) / (1 - (e**(-(V + 38) / 6)))
    mBeta = -(0.124 * (V + 38)) / (1 - (e**((V + 38) / 6)))
    mTau = 6 * (1 / (mAlpha + mBeta)) / QT
    
    # more voltage shifts
    if V == -17 or V == -64.4:
        V += 0.001

    # steady state inactivation and inactivation time constants (ms)
    hInf = 1.0 / (1 + e**((V + 48.8) / 10))
    hAlpha = -2.88e-6 * (V + 17) / (1 - e**((V + 17) / 4.63))
    hBeta = 6.94e-6 * (V + 64.4) / (1 - e**(-(V + 64.4) / 2.63))
    hTau = (1 / (hAlpha + hBeta)) / QT

    m, h = state

    dm_dt = (mInf - m) / mTau
    dh_dt = (hInf - h) / hTau

    return dm_dt, dh_dt

def current(state, V, E_Na):
    """
    Compute current based on gating variables and membrane potential.

    Inputs:
        - state (tuple of floats): (m, h) gating variables
        - V (float): membrane potential (mV)
        - E_Na (float): sodium reversal potential (mV)

    Returns: 
        - I_Na (float): current (mA/cm^2)
    """
    m, h = state
    
    gNap_Et2 = gNap_Et2bar * m * m * m * h
    I_Na = gNap_Et2 * (V - E_Na)
    return I_Na

File: src/cpp/Nap_Et2.cpp
#include "Nap_Et2.hpp"

#include <cmath>

namespace {
const scalar_t QT = std::pow(2.3, (34.0 - 21.0) / 10.0);
const scalar_t e = 2.718281828459045235360287471352;
const scalar_t gNap_Et2bar = 0.00001;
}

std::pair<scalar_t, scalar_t> Nap_Et2_derivatives(std::pair<scalar_t, scalar_t> state, scalar_t V) {
    const scalar_t mInf = 1.0 / (1.0 + std::pow(e, ((V + 52.6) / -4.6)));

    if (V == -38) {
        V += 0.0001;
    }

    const scalar_t mAlpha = (0.182 * (V + 38)) / (1 - (std::pow(e, (-(V + 38) / 6))));
    const scalar_t mBeta = -(0.124 * (V + 38)) / (1 - (std::pow(e, ((V + 38) / 6))));
    const scalar_t mTau = 6.0 * (1.0 / (mAlpha + mBeta)) / QT;

    if (V == -17 || V == -64.4) {
        V += 0.001;
    }

    const scalar_t hInf = 1.0 / (1.0 + std::pow(e, ((V + 48.8) / 10.0)));
    const scalar_t hAlpha = -2.88e-6 * (V + 17) / (1 - std::pow(e, ((V + 17) / 4.63)));
    const scalar_t hBeta = 6.94e-6 * (V + 64.4) / (1 - std::pow(e, (-(V + 64.4) / 2.63)));
    const scalar_t hTau = (1.0 / (hAlpha + hBeta)) / QT;

    const scalar_t m = state.first;
    const scalar_t h = state.second;

    const scalar_t dm_dt = (mInf - m) / mTau;
    const scalar_t dh_dt = (hInf - h) / hTau;

    return std::make_pair(dm_dt, dh_dt);
}

scalar_t Nap_Et2_current(std::pair<scalar_t, scalar_t> state, scalar_t V, scalar_t E_Na) {
    scalar_t m = state.first;
    scalar_t h = state.second;

    const scalar_t gNap_Et2 = gNap_Et2bar * m * m * m * h;
    const scalar_t I_Na = gNap_Et2 * (V - E_Na);
    return I_Na;
}